index_configuration_tool/tests/test_main.py

```diff
@@ -0,0 +1,172 @@
+    valid_plugin = (random.choice(main.SUPPORTED_ENDPOINTS), plugin_config)
+    config_section = copy.deepcopy(BASE_DATA)
+    config_section[TEST_KEY].append(valid_plugin)
+    return config_section
+
+
+class TestMain(unittest.TestCase):
+    # Run before each test
+    def setUp(self) -> None:
+        with open(LOGSTASH_EXPECTED_OUTPUT_FILE, "rb") as f:
+            self.loaded_logstash_config = pickle.load(f)
+
+    def test_get_auth(self):
+        # Empty input should produce empty auth tuple
+        self.assertFalse(main.get_auth(dict()))
+        base_data = {"user": "user", "password": "password"}
+        # Password without user should return empty auth tuple
+        test_data = dict(base_data)
+        del test_data["user"]
+        self.assertFalse(main.get_auth(test_data))
+        # User without password should return empty auth tuple
+        test_data = dict(base_data)
+        del test_data["password"]
+        self.assertFalse(main.get_auth(test_data))
+        # Test valid input
+        result = main.get_auth(base_data)
+        assert type(result) is tuple
+        self.assertEqual("user", result[0])
+        self.assertEqual("password", result[1])
+
+    def test_get_endpoint_info(self):
+        # Simple base case
+        test_config = create_plugin_config(["test"])
+        result = main.get_endpoint_info(test_config)
+        self.assertEqual("http://test/", result[0])
+        self.assertFalse(result[1])
+        # SSL enabled
+        test_config = create_plugin_config(["test"], True)
+        result = main.get_endpoint_info(test_config)
+        self.assertEqual("https://test/", result[0])
+        self.assertFalse(result[1])
+        # SSL disabled, invalid auth config
+        test_config = create_plugin_config(["test"], False, "test")
+        result = main.get_endpoint_info(test_config)
+        self.assertEqual("http://test/", result[0])
+        self.assertFalse(result[1])
+        # SSL disabled, valid auth config
+        test_config = create_plugin_config(["test"], user="user", password="password")
+        result = main.get_endpoint_info(test_config)
+        self.assertEqual("http://test/", result[0])
+        self.assertTrue(result[1])
+        # Array of hosts uses the first entry
+        test_config = create_plugin_config(["test1", "test2"], True, "user", "password")
+        result = main.get_endpoint_info(test_config)
+        self.assertEqual("https://test1/", result[0])
+        self.assertTrue(result[1])
+
+    def test_print_report(self):
+        # Base case should return an empty list
+        self.assertFalse(main.print_report(dict(), dict()))
+        base_data = {
+            "index1": {
+                "settings": {
+                    "bool_key": True,
+                    "num_key": 1
+                },
+                "mappings": {
+                    "str_key": "value"
+                }
+            },
+            "index2": {}
```

Comment:
what does it mean for an index to not have any settings or mappings? Is this a possible state?

Reply:
Yes, it's possible for an index to have no explicit mappings. Strictly speaking, the default minimum configuration (ignoring internal/transient entries) for an index is:
```
"mappings": {},
"settings": {
    "number_of_shards": 1,
    "number_of_replicas": 1
}
```

However, I think these default configurations are irrelevant for the purposes of the tests in this file. I'm totally fine to use this configuration if you feel otherwise though.